solidity/ecdsa/contracts/libraries/EcdsaAuthorization.sol

```diff
@@ -15,25 +15,474 @@
+        );
+        require(
+            decrease.decreasingAt != type(uint64).max,
+            "Authorization decrease request not activated"
+        );
+        require(
+            // solhint-disable-next-line not-rely-on-time
+            block.timestamp > decrease.decreasingAt,
+            "Authorization decrease delay not passsed"
+        );
+
+        emit AuthorizationDecreaseApproved(stakingProvider);
+
+        // slither-disable-next-line unused-return
+        tokenStaking.approveAuthorizationDecrease(stakingProvider);
+        delete self.pendingDecreases[stakingProvider];
+    }
+
+    /// @notice Used by T staking contract to inform the application the
+    ///         authorization has been decreased for the given staking provider
+    ///         involuntarily, as a result of slashing.
+    ///
+    ///         If the operator is not known (`registerOperator` was not called)
+    ///         the function does nothing. The operator was never in a sortition
+    ///         pool so there is nothing to update.
+    ///
+    ///         If the operator is known, sortition pool is unlocked, and the
+    ///         operator is in the sortition pool, the sortition pool state is
+    ///         updated. If the sortition pool is locked, update needs to be
+    ///         postponed. Every other staker is incentivized to call
+    ///         `updateOperatorStatus` for the problematic operator to increase
+    ///         their own rewards in the pool.
+    ///
+    /// @dev Should only be callable by T staking contract.
+    function involuntaryAuthorizationDecrease(
+        Data storage self,
+        IStaking tokenStaking,
+        SortitionPool sortitionPool,
+        address stakingProvider
+    ) internal {
+        address operator = self.stakingProviderToOperator[stakingProvider];
+
+        if (operator == address(0)) {
+            // Operator is not known. It means `registerOperator` was not
+            // called yet, and there is no chance the operator could
+            // call `joinSortitionPool`. We can just ignore this update because
+            // operator was never in the sortition pool.
+            return;
+        } else {
+            // Operator is known. It means that this operator is or was in the
+            // sortition pool and the sortition pool may need to be updated.
+            //
+            // If the sortition pool is not locked and the operator is in the
+            // sortition pool, we are updating it.
+            //
+            // If the sortition pool is locked, update needs to wait. Other
+            // sortition pool members are incentivized to call
+            // `updateOperatorStatus` for the problematic operator because they
+            // will increase their rewards this way.
+            if (
+                sortitionPool.isOperatorInPool(operator) &&
+                !sortitionPool.isLocked()
+            ) {
+                updateOperatorStatus(
```

Comment:
This is a result of how slashing works in T `TokenStaking`.

To keep stakes synchronized between applications when staking providers are slashed, without the risk of running out of gas, the staking contract queues up slashings and let users process the transactions. When an application slashes one or more staking providers, it [adds them to the slashing queue on the staking contract](https://github.com/threshold-network/solidity-contracts/blob/702aa5f12242e885e7b0c2b3228bb788439d30a8/contracts/staking/TokenStaking.sol#L1474-L1480). A queue entry contains the staking provider’s address and the amount they are due to be slashed.

When there is at least one staking provider in the slashing queue, any account can submit [a transaction processing one or more staking providers' slashings](https://github.com/threshold-network/solidity-contracts/blob/702aa5f12242e885e7b0c2b3228bb788439d30a8/contracts/staking/TokenStaking.sol#L1137), and collecting a reward for doing so. A queued slashing is processed by updating the staking provider’s stake to the post-slashing amount, updating authorized amount for each affected application, and notifying all affected applications that the staking provider’s authorized stake has been reduced due to slashing.

The entire idea is that the process transaction is expensive because each application needs to be updated, so the reward for the processor is hefty and comes from the slashed tokens. Practically, it means that if the sortition pool is unlocked, and can be updated, it should be updated because we already paid someone for updating it.

Reply:
Thanks for the clarification! Could we add a short comment in the code?

Reply:
Sure! See 9e6c3cbd80bf7f3e3a34e1173bb6eaf3553d57ab